core/org.wso2.carbon.user.core/src/main/java/org/wso2/carbon/user/core/common/AbstractUserStoreManager.java

```diff
@@ -8418,47 +8418,61 @@ public void addSystemRole(String roleName, String[] userList, Permission[] permi
 
     private List<String> getUserRolesWithID(String userID, String filter) throws UserStoreException {
 
-        List<String> internalRoles = doGetInternalRoleListOfUserWithID(userID, filter);
-        Set<String> modifiedInternalRoles = new HashSet<>();
-        String[] modifiedExternalRoleList = new String[0];
-
-        if (readGroupsEnabled && doCheckExistingUserWithID(userID)) {
-            List<String> roles = new ArrayList<>();
-            String[] externalRoles = doGetExternalRoleListOfUserWithID(userID, "*");
-            roles.addAll(Arrays.asList(externalRoles));
-            if (isSharedGroupEnabled()) {
-                String[] sharedRoles = doGetSharedRoleListOfUserWithID(userID, null, "*");
-                if (sharedRoles != null) {
-                    roles.addAll(Arrays.asList(sharedRoles));
+        List<String> userRoleList;
+        String username = getUserNameFromUserID(userID);
+        synchronized (userID.intern()) {
+            if (username != null) {
+                String[] roleListOfUserFromCache = getRoleListOfUserFromCache(this.tenantId, username);
+                if (roleListOfUserFromCache != null) {
```

Comment:
Better to start `synchronized` block only if `if (roleListOfUserFromCache == null) {` right?

Reply:
Here the `if (roleListOfUserFromCache == null)` check is done inside the synchronized block for double-checked locking purpose. There is another `if (roleListOfUserFromCache == null)` check in the calling method right before this method is called.

Reply:
@DinikaSen There is a pattern to follow when using synchronise block. Always advisable to follow the pattern so that code and logic is simpler and less error prone

if (! checkCondition()  ) {
syncronized (lock) {
if (checkCondition() ) {exit;}